### PR TITLE
Feature Request: Support GIFs

### PR DESCRIPTION
**Pull request by RbBtSn0w**
That's was support GIFs feature for AFNetworking. #3843

### CODE REVIEW

File: AFNetworking/AFURLResponseSerialization.h [hunk 1]
```diff
@@ -239,7 +239,7 @@ NS_ASSUME_NONNULL_BEGIN
 @property (nonatomic, assign) CGFloat imageScale;
 
 /**
- Whether to automatically inflate response image data for compressed formats (such as PNG or JPEG). Enabling this can significantly improve drawing performance on iOS when used with `setCompletionBlockWithSuccess:failure:`, as it allows a bitmap representation to be constructed in the background rather than on the main thread. `YES` by default.
+ Whether to automatically inflate response image data for compressed formats (such as PNG or JPEG or GIFs). Enabling this can significantly improve drawing performance on iOS when used with `setCompletionBlockWithSuccess:failure:`, as it allows a bitmap representation to be constructed in the background rather than on the main thread. `YES` by default.
```
Comment (tewha) [inline AFNetworking/AFURLResponseSerialization.h]: Comment needs tuning, like "(such as PNG, JPEG or GIF)"

File: AFNetworking/AFURLResponseSerialization.m [hunk 2]
```diff
@@ -555,17 +556,50 @@ + (UIImage *)af_safeImageWithData:(NSData *)data {
     return [[UIImage alloc] initWithCGImage:[image CGImage] scale:scale orientation:image.imageOrientation];
 }
 
+static float AFFrameDurationFromSourceAtIndex(CGImageSourceRef source, NSUInteger index){
+    float frameDuration = 0.1f;
+    CFDictionaryRef cfFrameProperties = CGImageSourceCopyPropertiesAtIndex(source, index, nil);
```
Comment (tewha) [inline AFNetworking/AFURLResponseSerialization.m]: Documented as `__nullable`. Is the rest of this code safe if NULL is returned?

File: AFNetworking/AFURLResponseSerialization.m [hunk 3]
```diff
@@ -578,9 +612,36 @@ + (UIImage *)af_safeImageWithData:(NSData *)data {
...
+        CGImageSourceRef source = CGImageSourceCreateWithData((__bridge CFDataRef)data, NULL);
+        size_t count = CGImageSourceGetCount(source);
+        UIImage *gifImage;
+        
+        if (count > 1) {
+            NSMutableArray *images = [NSMutableArray array];
+            NSTimeInterval duration = 0.0f;
+            
+            for (size_t i = 0; i < count; i++) {
+                CGImageRef image = CGImageSourceCreateImageAtIndex(source, i, NULL);
```
Comment (tewha) [inline AFNetworking/AFURLResponseSerialization.m]: Result is `__nullable`.

File: Tests/Tests/AFImageDownloaderTests.m [hunk 4]
```diff
@@ -25,6 +25,7 @@
 @interface AFImageDownloaderTests : AFTestCase
 @property (nonatomic, strong) NSURLRequest *pngRequest;
 @property (nonatomic, strong) NSURLRequest *jpegRequest;
+@property (nonatomic, strong) NSURLRequest *gifsRequest;
```
Comment (0xced) [inline Tests/Tests/AFImageDownloaderTests.m]: Why plural `gifsRequest` instead of `gifRequest`?

Comment (RbBtSn0w) [inline Tests/Tests/AFImageDownloaderTests.m]: Thank you. 🍻

File: AFNetworking/AFURLResponseSerialization.m (hunk 3 above)
Comment (RbBtSn0w) [inline AFNetworking/AFURLResponseSerialization.m]: Got it, Thanks bro. 🍻

**Comment by RbBtSn0w** [comment]
thanks, @0xced @tewha .
I had pull new request at #3843

**Comment by RbBtSn0w** [comment]
Commit at #3772 .

**Pull request closed without merging**